File: go/xml_test.go
package xml

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"

	jsonic "github.com/jsonicjs/jsonic/go"
)

// specEntry represents one row of a TSV spec file.
type specEntry struct {
	File     string
	Line     int
	Name     string
	Input    string // Escape-decoded XML source.
	Expected string // Raw cell: JSON text, or "ERROR" / "ERROR:code".
	Opts     string // Raw JSON (may be empty).
}

// specDir returns the absolute path to the shared TSV spec directory.
func specDir() string {
	return filepath.Join("..", "test", "spec")
}

// loadSpec reads a TSV spec file into a slice of specEntry. Comment and
// blank lines are skipped. Escapes in the `input` column are decoded
// via unescapeInput; the `expected` and `opts` columns are left raw so
// JSON's own escape rules are honoured by the downstream JSON parser.
func loadSpec(t *testing.T, path string) []specEntry {
	t.Helper()
	f, err := os.Open(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	defer f.Close()

	var out []specEntry
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1<<20), 1<<20)
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		line := scanner.Text()
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		cols := strings.Split(line, "\t")
		if len(cols) < 3 {
			t.Fatalf("%s:%d: expected at least 3 tab-separated columns, got %d", path, lineNo, len(cols))
		}
		entry := specEntry{
			File:     filepath.Base(path),
			Line:     lineNo,
			Name:     cols[0],
			Input:    unescapeInput(cols[1]),
			Expected: cols[2],
		}
		if len(cols) >= 4 {
			entry.Opts = cols[3]
		}
		out = append(out, entry)
	}
	if err := scanner.Err(); err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	return out
}

// unescapeInput decodes the escape sequences used in the `input`
// column of the TSV spec: \n (LF), \r (CR), \t (TAB), \\ (backslash).
// Any other `\x` sequence is left intact so XML escapes like `\d` are
// not accidentally rewritten.
func unescapeInput(s string) string {
	if !strings.Contains(s, `\`) {
		return s
	}
	var b strings.Builder
	b.Grow(len(s))
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '\\' && i+1 < len(s) {
			switch s[i+1] {
			case 'n':
				b.WriteByte('\n')
				i++
				continue
			case 'r':
				b.WriteByte('\r')
				i++
				continue
			case 't':
				b.WriteByte('\t')
				i++
				continue
			case '\\':
				b.WriteByte('\\')
				i++
				continue
			}
		}
		b.WriteByte(c)
	}
	return b.String()
}

// parseOpts decodes the optional options JSON into a map suitable for
// jsonic.UseDefaults. Empty strings produce an empty map.
func parseOpts(t *testing.T, entry specEntry) map[string]any {
	t.Helper()
	if strings.TrimSpace(entry.Opts) == "" {
		return map[string]any{}
	}
	var out map[string]any
	if err := json.Unmarshal([]byte(entry.Opts), &out); err != nil {
		t.Fatalf("%s:%d: parse opts %q: %v", entry.File, entry.Line, entry.Opts, err)
	}
	return out
}

// parseExpected decodes the expected cell: either a JSON document or
// an `ERROR` / `ERROR:code` marker.
func parseExpected(t *testing.T, entry specEntry) (wantErr bool, errCode string, wantJSON any) {
	t.Helper()
	raw := entry.Expected
	if strings.HasPrefix(raw, "ERROR") {
		rest := strings.TrimPrefix(raw, "ERROR")
		rest = strings.TrimPrefix(rest, ":")
		return true, rest, nil
	}
	if err := json.Unmarshal([]byte(raw), &wantJSON); err != nil {
		t.Fatalf("%s:%d: parse expected JSON %q: %v", entry.File, entry.Line, raw, err)
	}
	return false, "", wantJSON
}

// runSpecFile is the workhorse: it loads one spec file and runs each
// row as its own sub-test.
func runSpecFile(t *testing.T, path string) {
	entries := loadSpec(t, path)
	if len(entries) == 0 {
		t.Fatalf("%s: no spec entries loaded", path)
	}
	for _, entry := range entries {
		entry := entry
		t.Run(entry.Name, func(t *testing.T) {
			opts := parseOpts(t, entry)
			wantErr, errCode, wantVal := parseExpected(t, entry)

			j := jsonic.Make()
			if err := j.UseDefaults(Xml, Defaults, opts); err != nil {
				t.Fatalf("plugin init: %v", err)
			}
			got, err := j.Parse(entry.Input)

			if wantErr {
				if err == nil {
					t.Fatalf("expected parse error, got result %v", got)
				}
				if errCode != "" && !strings.Contains(err.Error(), errCode) {
					t.Fatalf("expected error code %q, got %q", errCode, err.Error())
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected parse error: %v", err)
			}

			// Round-trip the got value through JSON for type normalisation
			// so `[]any` vs concrete slice types compare cleanly against
			// values decoded from the spec via json.Unmarshal.
			gotJSON, err := json.Marshal(got)
			if err != nil {
				t.Fatalf("marshal got: %v", err)
			}
			var gotVal any
			if err := json.Unmarshal(gotJSON, &gotVal); err != nil {
				t.Fatalf("unmarshal got: %v", err)
			}
			if !reflect.DeepEqual(gotVal, wantVal) {
				wantPretty, _ := json.Marshal(wantVal)
				t.Fatalf("\nwant: %s\ngot : %s", wantPretty, gotJSON)
			}
		})
	}
}

func TestBasicSpec(t *testing.T)      { runSpecFile(t, filepath.Join(specDir(), "basic.tsv")) }
func TestAttributesSpec(t *testing.T) { runSpecFile(t, filepath.Join(specDir(), "attributes.tsv")) }
func TestEntitiesSpec(t *testing.T)   { runSpecFile(t, filepath.Join(specDir(), "entities.tsv")) }
func TestNamespacesSpec(t *testing.T) { runSpecFile(t, filepath.Join(specDir(), "namespaces.tsv")) }
func TestStructureSpec(t *testing.T)  { runSpecFile(t, filepath.Join(specDir(), "structure.tsv")) }
func TestErrorsSpec(t *testing.T)     { runSpecFile(t, filepath.Join(specDir(), "errors.tsv")) }
func TestW3CSpec(t *testing.T)        { runSpecFile(t, filepath.Join(specDir(), "w3c.tsv")) }

// --- XML embedded in Jsonic source -----------------------------------------
//
// Real-world use case: a Jsonic document holds an XML payload as a string.
// Parse the outer document with stock Jsonic, then feed the embedded XML
// string into a second Jsonic instance configured with the Xml plugin.

func TestXmlEmbeddedInJsonic(t *testing.T) {
	// An ordinary Jsonic document. Uses backtick-delimited multiline
	// strings so the XML can embed newlines and double quotes verbatim.
	jsonicSrc := "{\n" +
		"  title: 'order-42',\n" +
		"  payload: `" +
		`<?xml version="1.0"?>` + "\n" +
		`<order id="42">` + "\n" +
		`  <item qty="2">Widget</item>` + "\n" +
		`  <item qty="1">Gadget</item>` + "\n" +
		`</order>` + "`,\n" +
		"}\n"

	outer, err := jsonic.Parse(jsonicSrc)
	if err != nil {
		t.Fatalf("parse outer Jsonic: %v", err)
	}
	m, ok := outer.(map[string]any)
	if !ok {
		t.Fatalf("outer should be map, got %T", outer)
	}
	if m["title"] != "order-42" {
		t.Fatalf("title mismatch: %v", m["title"])
	}
	payload, ok := m["payload"].(string)
	if !ok {
		t.Fatalf("payload should be string, got %T", m["payload"])
	}

	// Parse the XML payload with the Xml plugin.
	xmlParser := jsonic.Make()
	if err := xmlParser.UseDefaults(Xml, Defaults); err != nil {
		t.Fatalf("xml plugin init: %v", err)
	}
	parsed, err := xmlParser.Parse(payload)
	if err != nil {
		t.Fatalf("parse XML payload: %v", err)
	}
	el, ok := parsed.(map[string]any)
	if !ok {
		t.Fatalf("xml result should be map, got %T", parsed)
	}
	if el["name"] != "order" {
		t.Fatalf("root name: got %v, want order", el["name"])
	}
	attrs, _ := el["attributes"].(map[string]any)
	if attrs["id"] != "42" {
		t.Fatalf("root attr id: got %v, want 42", attrs["id"])
	}
	// Count <item> children and check attrs.
	children, _ := el["children"].([]any)
	var items []map[string]any
	for _, c := range children {
		if cm, ok := c.(map[string]any); ok && cm["name"] == "item" {
			items = append(items, cm)
		}
	}
	if len(items) != 2 {
		t.Fatalf("expected 2 item elements, got %d", len(items))
	}
	if a, _ := items[0]["attributes"].(map[string]any); a["qty"] != "2" {
		t.Fatalf("item[0].qty: got %v, want 2", a["qty"])
	}
	if a, _ := items[1]["attributes"].(map[string]any); a["qty"] != "1" {
		t.Fatalf("item[1].qty: got %v, want 1", a["qty"])
	}
}

// TestSpecDirExists is a sanity check that the shared test/spec folder is
// reachable from the Go test working directory.
func TestSpecDirExists(t *testing.T) {
	info, err := os.Stat(specDir())
	if err != nil {
		t.Fatalf("spec dir: %v", err)
	}
	if !info.IsDir() {
		t.Fatalf("%s is not a directory", specDir())
	}
}

// Compile-time assertion that specEntry stringifies meaningfully in
// error messages (keeps `fmt` import stable if trimmed elsewhere).
var _ = fmt.Sprintf

File: test/xml.test.ts
/* Copyright (c) 2021-2025 Richard Rodger and other contributors, MIT License */

import { describe, test } from 'node:test'
import assert from 'node:assert'
import { readFileSync, readdirSync } from 'node:fs'
import { join } from 'node:path'

import { Jsonic } from 'jsonic'
import { Xml } from '../dist/xml'

// ---------------------------------------------------------------------------
// Shared TSV spec runner
//
// Test cases are defined in tab-separated value files under test/spec/*.tsv.
// Each non-comment row is:
//   name<TAB>input<TAB>expected<TAB>opts
// - `input` uses the escape set \n \r \t \\
// - `expected` is raw JSON (standard JSON escapes apply) or the literal
//   token ERROR / ERROR:code for expected parse failures.
// - `opts` is optional JSON for plugin options.
// The same files drive the Go test suite in go/xml_test.go.
// ---------------------------------------------------------------------------

// At runtime this test file is loaded from `dist-test/`, so hop up one
// level to reach the shared spec directory in the project root.
const specDir = join(__dirname, '..', 'test', 'spec')

type SpecRow = {
  file: string
  line: number
  name: string
  input: string
  expected: string
  opts: string
}

function loadSpec(file: string): SpecRow[] {
  const path = join(specDir, file)
  const body = readFileSync(path, 'utf8')
  const rows: SpecRow[] = []
  const lines = body.split('\n')
  for (let i = 0; i < lines.length; i++) {
    const raw = lines[i]
    if (raw === '' || raw.startsWith('#')) continue
    const cols = raw.split('\t')
    if (cols.length < 3) {
      throw new Error(`${file}:${i + 1}: expected >=3 tab-separated columns`)
    }
    rows.push({
      file,
      line: i + 1,
      name: cols[0],
      input: unescapeInput(cols[1]),
      expected: cols[2],
      opts: cols[3] ?? '',
    })
  }
  return rows
}

// Decode the escape sequences used in the spec `input` column. Keeps
// the behaviour identical to the Go loader so the two language test
// suites exercise the exact same XML text.
function unescapeInput(s: string): string {
  if (!s.includes('\\')) return s
  let out = ''
  for (let i = 0; i < s.length; i++) {
    const c = s[i]
    if (c === '\\' && i + 1 < s.length) {
      const n = s[i + 1]
      if (n === 'n') { out += '\n'; i++; continue }
      if (n === 'r') { out += '\r'; i++; continue }
      if (n === 't') { out += '\t'; i++; continue }
      if (n === '\\') { out += '\\'; i++; continue }
    }
    out += c
  }
  return out
}

function runSpec(file: string) {
  const rows = loadSpec(file)
  describe(file, () => {
    for (const row of rows) {
      test(row.name, () => {
        const opts = row.opts.trim() === '' ? undefined : JSON.parse(row.opts)
        const jx = opts ? Jsonic.make().use(Xml, opts) : Jsonic.make().use(Xml)

        if (row.expected.startsWith('ERROR')) {
          const code = row.expected.slice(5).replace(/^:/, '')
          assert.throws(
            () => jx(row.input),
            (err: Error) =>
              code === '' || err.message.includes(code) ||
              // Jsonic wraps codes as `jsonic/<code>`; accept that form too.
              err.message.includes('/' + code),
            `${row.file}:${row.line}: expected error ${row.expected}`,
          )
          return
        }

        const got = jx(row.input)
        const want = JSON.parse(row.expected)
        // Round-trip `got` through JSON so ordering of keys does not affect
        // structural comparison (deepEqual is already order-insensitive for
        // objects, but this also strips undefined fields cleanly).
        assert.deepEqual(
          JSON.parse(JSON.stringify(got)),
          want,
          `${row.file}:${row.line}: ${row.name}`,
        )
      })
    }
  })
}

// Auto-discover every .tsv under test/spec and run it. Keeping this
// driven by directory contents means adding a new spec file never
// requires editing the TypeScript test code.
for (const file of readdirSync(specDir)) {
  if (file.endsWith('.tsv')) runSpec(file)
}


// ---------------------------------------------------------------------------
// XML embedded in Jsonic source
//
// A common real-world pattern is to keep XML payloads inside a larger
// Jsonic configuration file as a string value. This test demonstrates
// that the stock Jsonic parser reads the outer document and the Xml
// plugin parses the embedded payload.
// ---------------------------------------------------------------------------

describe('xml-embedded-in-jsonic', () => {
  test('parses XML inside a Jsonic multiline string', () => {
    // A plain Jsonic document. The backtick string carries the XML
    // payload verbatim, with newlines and double quotes intact.
    const jsonicSrc = "{\n" +
      "  title: 'order-42',\n" +
      "  payload: `" +
      '<?xml version="1.0"?>\n' +
      '<order id="42">\n' +
      '  <item qty="2">Widget</item>\n' +
      '  <item qty="1">Gadget</item>\n' +
      '</order>' + "`,\n" +
      "}\n"

    const outer = Jsonic(jsonicSrc) as any
    assert.equal(outer.title, 'order-42')
    assert.equal(typeof outer.payload, 'string')

    const xmlParser = Jsonic.make().use(Xml)
    const parsed = xmlParser(outer.payload) as any
    assert.equal(parsed.name, 'order')
    assert.equal(parsed.attributes.id, '42')

    const items = parsed.children.filter(
      (c: any) => typeof c === 'object' && c.name === 'item',
    )
    assert.equal(items.length, 2)
    assert.equal(items[0].attributes.qty, '2')
    assert.equal(items[0].children[0], 'Widget')
    assert.equal(items[1].attributes.qty, '1')
    assert.equal(items[1].children[0], 'Gadget')
  })
})
